File: Code Python/03. Compression/03.2 Compression with FFT - Fast Fourier Transformation/AudioCompressionFFT.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.image import imread
import sounddevice as sd
import time
import soundfile as sf

audio, FS = sf.read('Für-Elise.wav')
audio = audio[0:300000, 0]

FC = np.fft.fft(audio)

perc = [0.99, 0.3, 0.1, 0.01, 0.001]

FCsort = abs(FC)
FCsort[::-1].sort()

plt.figure(figsize=(20, 10))

for i in range(0, len(perc)):
    limit = FCsort[int(np.round(perc[i] * len(FCsort)))]
    ind = abs(FC) > limit
    FCred = FC * ind
    audiocomp = np.fft.ifft(FCred)

    sd.play(np.real(audiocomp), FS)

    plt.subplot(2, 5, i+1)
    plt.plot(np.arange(0, len(audio)), np.real(audiocomp))
    plt.axis([1, len(audio), -1, 1])
    plt.title('compressed to ' + str(perc[i] * 100) + ' %')

    plt.subplot(2, 5, i + 6)
    PS = abs(FCred)
    plt.semilogy(np.arange(0, len(audio)), np.abs(np.fft.fftshift(PS)))
    plt.ylim([1e-3, 1e+4])
    time.sleep(len(audio)/FS + .5)
plt.show()

plt.figure(figsize=(20, 10))
plt.subplot(2, 1, 1)
plt.imshow(imread('NotesFür-Elise.jpg'))
plt.title('Tones')

dF = FS / len(audio)
f = np.arange(-FS/2, FS/2, dF)

# Defining colors
c = [0, 0.4470, 0.7410]
r = [1, 0, 0]
b = [0, 0, 1]
g = [0, 1, 0]
y = [1, 1, 0]

plt.subplot(212)
plt.axis([0, 3000, 0, 2200])
plt.plot(329.427, 2099.4, color=r, marker='o', linewidth=3)
plt.plot(439.677, 1119.72, color=b, marker='o', linewidth=3)
plt.plot(494.214, 1116.68, color=g, marker='o', linewidth=3)
plt.plot(621.81, 1091.37, color=y, marker='o', linewidth=3)
plt.plot(f, abs(np.fft.fftshift(FC)), color=c)
plt.plot(np.ones(2200)*329.6, np.arange(1, 2201), color='k')
plt.plot(np.ones(2200)*392, np.arange(1, 2201), color='k')
plt.plot(np.ones(2200)*493.8, np.arange(1, 2201), color='k')
plt.plot(np.ones(2200)*587.3, np.arange(1, 2201), color='k')
plt.plot(np.ones(2200)*698.5, np.arange(1, 2201), color='k')
plt.xlabel('frequencies in Hz')
plt.ylabel('|f| Power of frequency')
plt.title('Power Spectrum')
plt.legend(['E4', 'A4', 'B4', 'Dis5', 'Power Spectrum'])
plt.show()

File: Code Matlab/03. Compression/03.2 Compression with FFT - Fast Fourier Transformation/AudioCompressionFFT.m
clc, clear, close all
%% Import Data
[audio, FS] = audioread('Für-Elise.mp3');
audio = audio(1:300000,1);

%% FFT Data
FC = fft(audio);

%% IFFT Data
perc = [.99 .3 .1 .01 .001];

[FCsort, ind] = sort(FC,'descend');

figure('Name','Compressed Data')
c = [0 0.4470 0.7410];
for i = 1:length(perc)
    limit = abs(FCsort(round(perc(i)*length(audio))));
    clr = abs(FC)>limit;
    FCred = FC.*clr;
    
    audiocomp = ifft(FCred);
    
    %sound(audiocomp,FS)
    
    subplot(2,5,i)
    plot(1:length(audio),real(audiocomp))
    axis([1 length(audio) -1 1])
    title(['compressed to ', num2str(perc(i)*100),' %'])
    
    subplot(2,5,i+5)
    bar(abs(fftshift(FCred)),'FaceColor',c,'EdgeColor',c)
    set(gca,'YScale','log')
    xlim([1 length(audio)])
    ylim([1e-3 1e+4])
    %pause(length(audio)/FS + 0.5)
end

%% Correlation Notes and Frequencies

figure('Name','Correlation Notes and Frequencies')
subplot(2,1,1)
imshow(imread('NotesFür-Elise.jpg'))
title('Tones')

dF = FS/length(audio);                      % hertz
f = -FS/2:dF:FS/2-dF;

subplot(2,1,2)
hold on
axis([0 3000 0 2200])
plot(329.427,2099.4,'or','LineWidth',3)
plot(439.677,1119.72,'ob','LineWidth',3)
plot(494.214,1116.68,'og','LineWidth',3)
plot(621.81,1091.37,'oy','LineWidth',3)
plot(f,abs(fftshift(FC)),'LineWidth',2,'Color',c)
plot(ones(1,2200)*329.6,1:2200,'k')
plot(ones(1,2200)*392,1:2200,'k')
plot(ones(1,2200)*493.8,1:2200,'k')
plot(ones(1,2200)*587.3,1:2200,'k')
plot(ones(1,2200)*698.5,1:2200,'k')
xlabel('frequencies in Hz')
ylabel('|f| Power of frequency')
title('Power Spectrum')
legend('E4','A4','B4','Dis5','Power Spectrum')
